auth_totp/controllers/main.py

```diff
@@ -0,0 +1,124 @@
+
+    @http.route('/auth_totp/login', type='http', auth='none', methods=['POST'])
+    def mfa_login(self, *args, **kwargs):
+        token = request.params.get('mfa_login_token')
+        try:
+            user = request.env['res.users'].user_from_mfa_login_token(token)
+        except (MfaTokenInvalidError, MfaTokenExpiredError) as exception:
+            if isinstance(exception, MfaTokenInvalidError):
+                msg = 'Your MFA login token is not valid. Please try again.'
+            if isinstance(exception, MfaTokenExpiredError):
+                msg = 'Your MFA login token has expired. Please try again.'
+
+            return http.local_redirect(
+                '/web/login',
+                query={
+                    'redirect': request.params.get('redirect'),
+                    'error': msg,
+                },
+                keep_hash=True,
+            )
+
+        confirmation_code = request.params.get('confirmation_code')
+        if not user.sudo().validate_mfa_confirmation_code(confirmation_code):
+            return http.local_redirect(
+                '/auth_totp/login',
+                query={
+                    'redirect': request.params.get('redirect'),
+                    'error': 'Your confirmation code is not correct. Please'
+                             ' try again.',
+                    'mfa_login_token': token,
+                },
+                keep_hash=True,
+            )
+
+        # These context managers trigger a safe commit, which persists the
+        # changes right away and is needed for the auth call
+        with Environment.manage():
+            with registry(request.db).cursor() as temp_cr:
+                temp_env = Environment(temp_cr, SUPERUSER_ID, request.context)
+                temp_user = temp_env['res.users'].browse(user.id)
+                temp_user.generate_mfa_login_token(60 * 24 * 30)
+                token = temp_user.mfa_login_token
+        request.session.authenticate(request.db, user.login, token, user.id)
+
+        redirect = request.params.get('redirect')
+        if not redirect:
+            redirect = '/web'
+        response = Response(http.redirect_with_hash(redirect))
+
+        if request.params.get('remember_device'):
+            device_model = request.env['res.users.device']
+            device = device_model.sudo().create({'user_id': user.id})
+            config_model = request.env['ir.config_parameter'].sudo()
+            secret = config_model.get_param('database.secret')
+            device_cookie = JsonSecureCookie({'device_id': device.id}, secret)
+            cookie_lifetime = timedelta(days=30)
+            cookie_exp = datetime.now() + cookie_lifetime
+            device_cookie = device_cookie.serialize(cookie_exp)
+            cookie_key = 'trusted_devices_%s' % user.id
+            cookie_security = config_model.get_param('auth_totp.secure_cookie')
+            security_flag = False if cookie_security == '0' else True
```

Comment:
This is true; if `cookie_security` isn't an str of a number you'd run into issues.

I hate ternary because of PHP. See:

![Horse](http://i.imgur.com/1zgFd.jpg)